### PR TITLE
users reducer: Handle all user updates from realm_user op: update

### PR DESCRIPTION
**Pull request by chrisbobbe**
And, to get there (and also just because it's good practice), write down with a Flow type what exactly we expect to get from servers in this event.

In particular, this is on the path to correctly resolving #5250, as well as #2900 / #5363, which are server release goals.

Replaces: #3403
Related: #3408

### CODE REVIEW

File: src/users/__tests__/usersReducer-test.js [hunk 1]
```diff
@@ -51,22 +52,15 @@ describe('usersReducer', () => {
...
     // Tell ESLint to recognize `check` as a helper function that runs
     // assertions.
     /* eslint jest/expect-expect: ["error", { "assertFunctionNames": ["expect", "check"] }] */
-    const check = (personMaybeWithoutId: $Rest<User, { ... }>) => {
-      const person = {
-        user_id: theUser.user_id,
-        ...personMaybeWithoutId,
-      };
+    // TODO: Try to make `user_id` optional again.
+    const check = (person: $PropertyType<RealmUserUpdateEvent, 'person'>) => {
```
Comment (gnprice) [inline src/users/__tests__/usersReducer-test.js]: I have a fix for this, which I showed you in the office. Here's [a try-flow link](https://flow.org/try/#0C4TwDgpgBAqlC8AoKAfKBvKAPAXFAzsAE4CWAdgOYA0UARnmQK4C2tERUAvsmprgcXLUoAYwYs2HTgG5EiUJCgB5BBigBqRnjgy5Aej1QAKgAsS+KAHcA9kQDW+HIgAUW2AEoEAPgzJR1skJsbVVMADoIxi5ZbkQDOkZgKxMIMigSYAByCzYzMgATKABDNIASAFEAGwhmVOAjcGgAWibkkhETAhNrRkrCtigIAEdGEgA3IuqyYCcFCHxykfHJuoAeCura6YbIVaUaTMZMrxoYL2d3WXiWqGAUoghs4rT2IlsnVzwNmrqdiD2DkcvJ54D50H4RAEgvxvlt6o0AVBDsdQlAImEorpYvFTNBXrYoCYimBIIECJARCQAGbtSaVEDJVK3FJQWG-RrpCzkOKGO7QO5CKyPB7kh5FfJCD5uNnbBH7JFAkFgiFQpL8ODwNTozExfSGEqFfLWeZkTJJIkkpmWFJpDJPAL09JpPnMwWWYXQKmVayWIQ85nQfBgMWFcjAaxS7RK3xQfyBNVfKo-WW7eXInya8KRaKIbGGACCVlsdiKb0YBU5t2sUGYRTs0CKtw5YFLRVqwHYzKKWRyPQKEH6IH9Lpl8MgNANUEY+H5RKS5EIEHFUGsVIDrKTcL+6WAM8qVKcqxgic27NTgOO5zcMGj4NjkPjwVgqO1ObzUGcAGUaiVgLTKo6GQEBANQWJUJD1l2ST4N0vT9A2LxEG8HDUgG-r4G2s6CtOgourQfaFHMUASg8IjAIBFglIMSG2GEsYAEKJP69wQDQYbsGA1j4Lhc5NooLZEJhHYcDOoFVl0JAQH0aHxEGIZQFSbzMMk3YQGMna0IwFCVkUVJUhAZGCqOfxhO4cjyI0CxLBMUzAOsAAKbyQEQoB-KsmCMAA-CEnAXicz5oGM1gkPk5yXBZkBWaMNlrKUjnWM5rkIh5Pl+acYWyHMUXLLZ6ybme-wed5z6+QqxynKgUBBSFGURfMizRSs0x5aeKaFVOqVlf5ZwXLIiAQFgnEuQp5aGQEfH1dZTV2UYpx4LNUAAPoAJLTOwZCTAAItYABy1jAPZRT4I4z6akYYUYLE-oPkEHQGXYqjOP6sbOfgASLSeyZjv8ABKS6VMwMAzkQMBgPkqnlOp0wHK9ATlf6t7PXGQSw2Qn1bgif2TIDwOg+DHaQ3UMPsG9poZr48SxrG6Ko4tVBI7o8S6MOHL2STAQAMLWMwzDjUglOVZ+jC0DOB1Oewf7zEewNKEQ9H7TQ6BoJowOLSFeDMCQWADjQ6j4cAi0+mQ7Bq-kGtawOqCcF4SNoELIsQGLCUS5J+DS+wsvy8AivK9OJvqzWFv5Lr+RSeM7AgItP4kJU5va4UKDW7bGAoEjVMaH7RCm+IrDsPTAvp+oIjTuGzCLcGq4xxAi00lJZsCKQlD54Y6eJ8nSsZ6rAdMLnRC68blhR7WMd4IQjfaYnfXxANQ1JERWMA0D7B4xDUNJJmqcC+oCV4IcYOqZkzdU+oqN4GzRCk1zPMBP6k-mdPg22HPHJL0Qnv7aor9XIYM9PxNsDA1CJvFund-b1x7pII+sZ1Ch3AupIgkdo6VGKmPIQUCND60NpYY2WcQrFQgZ2NATAAJHzvnqQYj9hpEXtqLeKiVXarAWmcVQpQlC0AAFYAFliQkCPDQVYABpGgAA1c4AjozGQRAtARXgvCyCAA) for the exploration we did of this Flow issue.

**Review by gnprice — COMMENTED** [review]
Thanks @chrisbobbe! This looks great; tiny comment below, and a workaround for that Flow issue.

**Comment by chrisbobbe** [comment]
Revision pushed! Please hold off merging until I've done some manual testing, but you can look at the revision in the meantime.

File: src/users/__tests__/usersReducer-test.js [hunk 2]
```diff
@@ -99,13 +99,18 @@ describe('usersReducer', () => {
       check({ timezone: randString() });
     });
 
-    // Excluded: "When the owner of a bot changes." The `users` state
-    // doesn't include cross-realm bots.
+    test.skip('When the owner of a bot changes.', () => {
+      check({ user_id: theUser.user_id, bot_owner_id: (theUser.bot_owner_id ?? 1) + 1 });
```
Comment (gnprice) [inline src/users/__tests__/usersReducer-test.js]: nit: this and the rest of this commit still introduce setting user_id, then that's cut in a later commit

File: src/api/transportTypes.js [hunk 3]
```diff
@@ -18,6 +18,9 @@
 export type Auth = $ReadOnly<{|
   realm: URL,
   apiKey: string,
+
+  // TODO: Follow changes in the user's email address, whether while we're
+  //   event polling or not.
   email: string,
```
Comment (gnprice) [inline src/api/transportTypes.js:24]: I also went and added a mention of this case on #3408.

Comment (chrisbobbe) [reply src/api/transportTypes.js:24]: Thanks!

**Comment by chrisbobbe** [comment]
Thanks! Revision pushed, with the results of my manual testing: https://github.com/zulip/zulip/pull/22103, and a commit at the tip that follows from that.

File: src/users/usersReducer.js [hunk 4]
```diff
@@ -32,10 +33,59 @@ export default (
...
+                            value: person.custom_profile_field.value,
+                            rendered_value: person.custom_profile_field.rendered_value,
+
+                            // This assertion is cumbersome. But it fills a
+                            // gap in Flow's coverage…that apparently Flow
+                            // doesn't announce by marking anything with
+                            // `any`. Remove when doing so doesn't stop Flow
+                            // from catching something wrong on
+                            // `value` or `rendered_value`.
+                          }: $Values<$NonMaybeType<$PropertyType<User, 'profile_data'>>>),
```
Comment (gnprice) [inline src/users/usersReducer.js:64]: nit: I like to mark situations like this with something like `FlowIssue` (no `$`) in the comment. That helps in finding it, either for a broad sweep for issues we've run into in Flow (perhaps to consider dropping the workarounds after a Flow upgrade occasionally?)… or for finding a specific issue one is thinking of but doesn't remember precisely where we encountered it.

**Review by gnprice — COMMENTED** [review]
Thanks! The revision looks good, with one nit below. Please merge at will.

**Comment by chrisbobbe** [comment]
> Thanks! The revision looks good, with one nit below. Please merge at will.

Thanks for the review! Merged, with that tweak.

**Comment by timabbott** [comment]
It's really exciting to see this!

**Comment by chrisbobbe** [comment]
Glad to hear it, Tim! You might also be excited to see that we've finally got round to #4933.

**Pull request merged**